torch/_higher_order_ops/auto_functionalize.py

```diff
@@ -365,7 +375,20 @@ def __call__(
 auto_functionalized_v2.fallthrough(DispatchKey.AutogradCUDA)
 
 
-def can_auto_functionalize(op: OperatorBase) -> bool:
+def can_auto_functionalize(op: Union[OperatorBase, HigherOrderOperator]) -> bool:
+    if isinstance(op, HigherOrderOperator):
+        # HOPs that implements gen_shema is auto_functionalizable.
+        # However, it's up to each hop's functioanlization implementation to decide
+        # whether do_auto_functionalized is called when input mutation happens.
+
+        def _has_gen_schema(op: HigherOrderOperator):
+            method = "gen_schema"
+            return hasattr(type(op), method) and getattr(
+                type(op), method
+            ) is not getattr(HigherOrderOperator, method)
+
+        return _has_gen_schema(op)
```

Comment:
can_auto_functionalize should return False if hop has a gen_schema and the schema is functional

Reply:
If the schema is functional, we shouldn't auto-functionalize -- that complicates the IR